File: lib/src/models/key_signature.dart
import 'package:meta/meta.dart';
import 'package:musicxml_parser/src/exceptions/musicxml_structure_exception.dart';
import 'package:musicxml_parser/src/exceptions/musicxml_validation_exception.dart';
import 'package:musicxml_parser/src/parser/xml_helper.dart';
import 'package:musicxml_parser/src/utils/validation_utils.dart';
import 'package:xml/xml.dart';

/// Represents a key signature in a musical score.
///
/// A key signature is defined by the number of sharps or flats (fifths)
/// and an optional mode (e.g., major, minor).
///
/// Objects of this class are immutable.
@immutable
class KeySignature {
  /// The number of sharps (positive value) or flats (negative value)
  /// in the key signature. Typically ranges from -7 to 7.
  final int fifths;

  /// The mode of the key signature (e.g., "major", "minor", "dorian").
  ///
  /// This is optional as per MusicXML specification.
  final String? mode;

  /// Creates a new [KeySignature] instance.
  ///
  /// It's recommended to use [KeySignature.validated] or [KeySignature.fromXmlElement]
  /// for creating instances, as they include validation.
  const KeySignature({
    required this.fifths,
    this.mode,
  });

  /// Creates a new [KeySignature] instance with validation.
  ///
  /// This factory constructor performs validation against MusicXML rules
  /// (e.g., fifths within range, valid mode if specified).
  ///
  /// Throws [MusicXmlValidationException] if the key signature is invalid.
  ///
  /// Parameters:
  ///   [fifths]: The number of sharps or flats.
  ///   [mode]: The mode of the key (optional).
  ///   [line]: The line number in the XML document (for context in error messages).
  ///   [context]: Additional context for error messages.
  factory KeySignature.validated({
    required int fifths,
    String? mode,
    int? line,
    Map<String, dynamic>? context,
  }) {
    final keySignature = KeySignature(fifths: fifths, mode: mode);
    // ValidationUtils.validateKeySignature will throw if invalid.
    ValidationUtils.validateKeySignature(keySignature,
        line: line, context: context);
    return keySignature;
  }

  /// Creates a new [KeySignature] instance from a MusicXML `<key>` [element].
  ///
  /// This factory parses the required `<fifths>` element and the optional
  /// `<mode>` element. It then validates the parsed values.
  ///
  /// Throws [MusicXmlStructureException] if required XML elements are missing.
  /// Throws [MusicXmlValidationException] if the parsed values are invalid.
  ///
  /// Parameters:
  ///   [element]: The XML element representing the `<key>`.
  ///   [partId]: The ID of the part (for context in error messages).
  ///   [measureNumber]: The number of the measure (for context in error messages).
  factory KeySignature.fromXmlElement(
    XmlElement element,
    String partId,
    String measureNumber,
  ) {
    final line = XmlHelper.getLineNumber(element);

    final fifthsElement = element.findElements('fifths').firstOrNull;
    if (fifthsElement == null) {
      throw MusicXmlStructureException(
        'Required <fifths> element not found in <key>',
        parentElement: 'key',
        line: line,
        context: {'part': partId, 'measure': measureNumber},
      );
    }
    final fifthsText = fifthsElement.innerText.trim();
    final fifths = int.tryParse(fifthsText);

    if (fifths == null) {
      throw MusicXmlValidationException(
        'Invalid key signature fifths value: "$fifthsText". Must be an integer.',
        rule: 'key_fifths_invalid',
        line: XmlHelper.getLineNumber(fifthsElement),
        context: {
          'part': partId,
          'measure': measureNumber,
          'parsedFifths': fifthsText
        },
      );
    }

    final modeElement = element.findElements('mode').firstOrNull;
    final mode = modeElement?.innerText.trim();

    // Use the .validated factory to ensure consistent validation logic
    return KeySignature.validated(
      fifths: fifths,
      mode: mode,
      line: line,
      context: {'part': partId, 'measure': measureNumber},
    );
  }


  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is KeySignature &&
          runtimeType == other.runtimeType &&
          fifths == other.fifths &&
          mode == other.mode;

  @override
  int get hashCode => fifths.hashCode ^ (mode?.hashCode ?? 0);

  @override
  String toString() => 'KeySignature{fifths: $fifths, mode: $mode}';
}

File: csharp/MusicXMLParser/Models/KeySignature.cs
using System;
using System.Collections.Generic;
using System.Xml.Linq;
using MusicXMLParser.Exceptions; // Assuming these will be created later
using MusicXMLParser.Utils; // Assuming these will be created later
using MusicXMLParser.Parser; // Assuming XmlHelper will be here

namespace MusicXMLParser.Models
{
    /// <summary>
    /// Represents a key signature in a musical score.
    /// A key signature is defined by the number of sharps or flats (fifths)
    /// and an optional mode (e.g., major, minor).
    /// Objects of this class are immutable.
    /// </summary>
    public class KeySignature : IEquatable<KeySignature>
    {
        /// <summary>
        /// The number of sharps (positive value) or flats (negative value)
        /// in the key signature. Typically ranges from -7 to 7.
        /// </summary>
        public int Fifths { get; }

        /// <summary>
        /// The mode of the key signature (e.g., "major", "minor", "dorian").
        /// This is optional as per MusicXML specification.
        /// </summary>
        public string Mode { get; }

        /// <summary>
        /// Creates a new <see cref="KeySignature"/> instance.
        /// It's recommended to use <see cref="Validated"/> or <see cref="FromXElement"/>
        /// for creating instances, as they include validation.
        /// </summary>
        public KeySignature(int fifths, string mode = null)
        {
            Fifths = fifths;
            Mode = mode;
        }

        /// <summary>
        /// Creates a new <see cref="KeySignature"/> instance with validation.
        /// This factory method performs validation against MusicXML rules
        /// (e.g., fifths within range, valid mode if specified).
        /// Throws <see cref="MusicXmlValidationException"/> if the key signature is invalid.
        /// </summary>
        public static KeySignature Validated(int fifths, string mode = null, int? line = null, Dictionary<string, object> context = null)
        {
            var keySignature = new KeySignature(fifths, mode);
            // ValidationUtils.ValidateKeySignature will throw if invalid.
            ValidationUtils.ValidateKeySignature(keySignature, line, context);
            return keySignature;
        }

        /// <summary>
        /// Creates a new <see cref="KeySignature"/> instance from a MusicXML <key> <paramref name="element"/>.
        /// This factory parses the required <fifths> element and the optional
        /// <mode> element. It then validates the parsed values.
        /// Throws <see cref="MusicXmlStructureException"/> if required XML elements are missing.
        /// Throws <see cref="MusicXmlValidationException"/> if the parsed values are invalid.
        /// </summary>
        public static KeySignature FromXElement(XElement element, string partId, string measureNumber)
        {
            var line = XmlHelper.GetLineNumber(element);

            var fifthsElement = element.Element("fifths");
            if (fifthsElement == null)
            {
                throw new MusicXmlStructureException(
                    "Required <fifths> element not found in <key>",
                    parentElement: "key",
                    line: line,
                    context: new Dictionary<string, object> { { "part", partId }, { "measure", measureNumber } }
                );
            }
            var fifthsText = fifthsElement.Value.Trim();
            if (!int.TryParse(fifthsText, out var fifths))
            {
                throw new MusicXmlValidationException(
                    $"Invalid key signature fifths value: \"{fifthsText}\". Must be an integer.",
                    rule: "key_fifths_invalid",
                    line: XmlHelper.GetLineNumber(fifthsElement),
                    context: new Dictionary<string, object>
                    {
                        { "part", partId },
                        { "measure", measureNumber },
                        { "parsedFifths", fifthsText }
                    }
                );
            }

            var modeElement = element.Element("mode");
            var mode = modeElement?.Value.Trim();

            // Use the .Validated factory to ensure consistent validation logic
            return Validated(
                fifths: fifths,
                mode: mode,
                line: line,
                context: new Dictionary<string, object> { { "part", partId }, { "measure", measureNumber } }
            );
        }

        public override bool Equals(object obj) => Equals(obj as KeySignature);

        public bool Equals(KeySignature other) =>
            other != null &&
            Fifths == other.Fifths &&
            Mode == other.Mode;

        public override int GetHashCode() => HashCode.Combine(Fifths, Mode);

        public override string ToString() => $"KeySignature{{fifths: {Fifths}, mode: {Mode}}}";
    }
}
